### DISCUSSION TITLE
Navigating Pages

### DISCUSSION THREAD

**Discussion by jjwb96** [microsoft/WindowsAppSDK#1258] [Q&A] [answered]
I have noticed since updating to 0.8 that when I have the following scenario the code Executes before it loads/displays the page. 

```
On my MainWindow.xaml.cs I have

rootFrame.Navigate(typeof(ApplySettings));

On my ApplySettings.xaml.cs after the page initializes I have the following code

Loaded += ApplySettings_Loaded;

private void ApplySettings_Loaded(object sender, RoutedEventArgs e)
{
   \\ Does UI stuff while apply settings
}

```
On 0.5 it would Navigate to the next page as normal then start to execute the code, but on 0.8 it freezes on the main window and doesn't load the next page until all the code has executed. I have a ProgressRing and TextBlock on my apply settings page so that the user knows whats happening. 

What's the proper way of having code launch once the next page is visible?

**Comment by andrewleader** [comment, marked as answer]
Operations in your Loaded callback are performed **on the UI thread** by default, which means any non-async operations will block the UI, which is likely why you're seeing the page freeze up till it finishes the operations.

What kind of operations are you doing where you said `\\Does UI stuff while apply settings`? If those operations involve loading data or other non-UI operations, you can likely try to offload those to a separate thread.

We'll need to see the rest of your code to help you more though.

**Reply by jjwb96** [reply]
Thanks for the Response Andrew!

Would something like this work?

https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/keep-the-ui-thread-responsive

**Reply by andrewleader** [reply]
Ah yep that's a great article on the topic 😊

**Reply by jjwb96** [reply]
Great Thank you! If I run into any trouble I will post my code 😊 Thanks Again!

**Reply by jjwb96** [reply]
That seemed to have done the trick! Thanks again!
